### PR TITLE
add validators to a subnet in a devnet

### PR DESCRIPTION
**Pull request by felipemadero**
Closes https://github.com/ava-labs/avalanche-cli/issues/1159

### CODE REVIEW

File: cmd/nodecmd/validatePrimary.go [hunk 1]
```diff
@@ -177,17 +167,7 @@ func joinAsPrimaryNetworkValidator(nodeID ids.NodeID, network models.Network, no
...
-		ids.Empty,
-		nodeID, weight,
-		uint64(start.Unix()),
-		uint64(start.Add(duration).Unix()),
-		recipientAddr,
-		delegationFee,
-		nil,
-		signer.NewProofOfPossession(blsSk),
-	)
+	_, err = deployer.AddPermissionlessValidator(ids.Empty, ids.Empty, nodeID, weight, uint64(start.Unix()), uint64(start.Add(duration).Unix()), recipientAddr, delegationFee, nil, signer.NewProofOfPossession(blsSk))
```
Comment (arturrez) [inline cmd/nodecmd/validatePrimary.go:170]: make it multiline

File: cmd/nodecmd/validatePrimary.go [hunk 2]
```diff
@@ -196,6 +176,9 @@ func PromptWeightPrimaryNetwork(network models.Network) (uint64, error) {
 	if network.Kind == models.Mainnet {
```
Comment (arturrez) [inline cmd/nodecmd/validatePrimary.go:176]: nit: combine with```	if network.Kind == models.Mainnet {
		delegationFee = genesis.MainnetParams.MinDelegationFee
	}
	if network.Kind == models.Devnet {
		delegationFee = genesis.LocalParams.MinDelegationFee
	}```

File: cmd/nodecmd/validatePrimary.go [hunk 3]
```diff
@@ -196,6 +176,9 @@ func PromptWeightPrimaryNetwork(network models.Network) (uint64, error) {
...
 		defaultStake = genesis.MainnetParams.MinValidatorStake
 	}
+	if network.Kind == models.Devnet {
```
Comment (arturrez) [inline cmd/nodecmd/validatePrimary.go:179]: same

File: cmd/nodecmd/validatePrimary.go [hunk 4]
```diff
@@ -372,7 +362,26 @@ func validatePrimaryNetwork(_ *cobra.Command, args []string) error {
...
+		keyName,
+		useEwoq,
+		&useLedger,
+		ledgerAddresses,
+	)
+	if err != nil {
+		return err
+	}
+
+	err = setupAnsible(clusterName)
```
Comment (arturrez) [inline cmd/nodecmd/validatePrimary.go:384]: nit: `if err:=...; err!=nil`

File: cmd/nodecmd/validatePrimary.go [hunk 5]
```diff
@@ -443,7 +452,7 @@ func convertNanoAvaxToAvaxString(weight uint64) string {
 
 func PrintNodeJoinPrimaryNetworkOutput(nodeID ids.NodeID, weight uint64, network models.Network, start time.Time) {
 	ux.Logger.PrintToUser("NodeID: %s", nodeID.String())
-	ux.Logger.PrintToUser("Network: %s", network.Name())
+	ux.Logger.PrintToUser("Network: %s", network.Kind.String())
```
Comment (arturrez) [inline cmd/nodecmd/validatePrimary.go:455]: use Name() :)

File: cmd/nodecmd/validateSubnet.go [hunk 6]
```diff
@@ -71,9 +80,16 @@ func parseSubnetSyncOutput(filePath string) (string, error) {
 	return "", errors.New("unable to parse subnet sync status")
 }
 
-func addNodeAsSubnetValidator(nodeID, subnetName string, network models.Network, currentNodeIndex, nodeCount int) error {
+func addNodeAsSubnetValidator(network models.Network, kc keychain.Keychain, useLedger bool, nodeID, subnetName string, currentNodeIndex, nodeCount int) error {
```
Comment (arturrez) [inline cmd/nodecmd/validateSubnet.go:83]: nit: multiline

File: cmd/nodecmd/validateSubnet.go [hunk 7]
```diff
@@ -124,10 +140,33 @@ func waitForNodeToBePrimaryNetworkValidator(nodeID ids.NodeID) error {
...
+		keyName,
+		useEwoq,
+		&useLedger,
+		ledgerAddresses,
+	)
+	if err != nil {
+		return err
+	}
+
+	err = setupAnsible(clusterName)
```
Comment (arturrez) [inline cmd/nodecmd/validateSubnet.go:169]: nit: single line

File: cmd/subnetcmd/addValidator.go [hunk 8]
```diff
@@ -49,98 +51,87 @@ Testnet or Mainnet.`,
...
-			if err != nil {
-				return err
-			}
-		}
-	case models.Mainnet:
-		useLedger = true
-		if keyName != "" {
-			return ErrStoredKeyOnMainnet
+	if outputTxPath != "" {
+		if _, err := os.Stat(outputTxPath); err == nil {
```
Comment (arturrez) [inline cmd/subnetcmd/addValidator.go:119]: use utils.FileExists()

File: cmd/subnetcmd/addValidator.go [hunk 9]
```diff
@@ -49,98 +51,87 @@ Testnet or Mainnet.`,
 		RunE:         addValidator,
 		Args:         cobra.ExactArgs(1),
 	}
-	cmd.Flags().StringVarP(&keyName, "key", "k", "", "select the key to use [fuji deploy only]")
+	cmd.Flags().StringVarP(&keyName, "key", "k", "", "select the key to use [fuji/devnet only]")
 	cmd.Flags().StringVar(&nodeIDStr, "nodeID", "", "set the NodeID of the validator to add")
 	cmd.Flags().Uint64Var(&weight, "weight", 0, "set the staking weight of the validator to add")
 	cmd.Flags().StringVar(&startTimeStr, "start-time", "", "UTC start time when this validator starts validating, in 'YYYY-MM-DD HH:MM:SS' format")
+	cmd.Flags().BoolVar(&defaultValidator, "default-validator", false, "use default weight/start/duration params for subnet validator")
```
[the remark this answers is not in the document]
Comment (felipemadero) [reply cmd/subnetcmd/addValidator.go:58]: using default-validator-params

File: cmd/subnetcmd/addValidator.go [hunk 10]
```diff
@@ -278,6 +269,20 @@ func getTimeParameters(network models.Network, nodeID ids.NodeID, isValidator bo
...
 	)
 
+	defaultStakingStartLeadTime := constants.StakingStartLeadTime
+	if network.Kind == models.Devnet {
+		defaultStakingStartLeadTime = constants.DevnetStakingStartLeadTime
+	}
+
+	if defaultValidator {
+		start = time.Now().Add(defaultStakingStartLeadTime)
+		duration, err = getMaxValidationTime(network, nodeID, start)
```
[the remark this answers is not in the document]
Comment (felipemadero) [reply cmd/subnetcmd/addValidator.go:278]: for subnet validators let's use the primary validation window

File: cmd/subnetcmd/addValidator.go [hunk 11]
```diff
@@ -360,6 +365,9 @@ func PromptNodeID() (ids.NodeID, error) {
 }
 
 func PromptWeight() (uint64, error) {
+	if defaultValidator {
+		return constants.DefaultStakeWeight, nil
```
[the remark this answers is not in the document]
Comment (felipemadero) [reply cmd/subnetcmd/addValidator.go:368]: not so for subnets

**Pull request merged**